src/audio/manager.cpp

```diff
@@ -130,70 +130,125 @@ AudioManager::AudioManager(const Glib::ustring &backends_string)
 
     Enumerate();
 
-    m_playback_config = ma_device_config_init(ma_device_type_playback);
-    m_playback_config.playback.format = ma_format_f32;
-    m_playback_config.playback.channels = 2;
-    m_playback_config.sampleRate = 48000;
-    m_playback_config.dataCallback = data_callback;
-    m_playback_config.pUserData = this;
-
-    if (const auto playback_id = m_devices.GetDefaultPlayback(); playback_id.has_value()) {
-        m_playback_id = *playback_id;
-        m_playback_config.playback.pDeviceID = &m_playback_id;
-
-        if (auto code = ma_device_init(&m_context, &m_playback_config, &m_playback_device); code != MA_SUCCESS) {
-            spdlog::get("audio")->error("failed to initialize playback device (code: {})", static_cast<int>(code));
-            m_ok = false;
-            return;
-        }
+    Glib::signal_timeout().connect(sigc::mem_fun(*this, &AudioManager::DecayVolumeMeters), 40);
+}
 
-        if (auto code = ma_device_start(&m_playback_device); code != MA_SUCCESS) {
-            spdlog::get("audio")->error("failed to start playback (code: {})", static_cast<int>(code));
-            ma_device_uninit(&m_playback_device);
-            m_ok = false;
-            return;
-        }
+AudioManager::~AudioManager() {
+    if (m_playback_device_ready) {
+        ClosePlaybackDevice();
+    }
 
-        char playback_device_name[MA_MAX_DEVICE_NAME_LENGTH + 1];
-        ma_device_get_name(&m_playback_device, ma_device_type_playback, playback_device_name, sizeof(playback_device_name), nullptr);
-        spdlog::get("audio")->info("using {} as playback device", playback_device_name);
+    if (m_capture_device_ready) {
+        CloseCaptureDevice();
     }
 
-    m_capture_config = ma_device_config_init(ma_device_type_capture);
-    m_capture_config.capture.format = ma_format_s16;
-    m_capture_config.capture.channels = 2;
-    m_capture_config.sampleRate = 48000;
-    m_capture_config.periodSizeInFrames = 480;
-    m_capture_config.dataCallback = capture_data_callback;
-    m_capture_config.pUserData = this;
+    ma_context_uninit(&m_context);
+    RemoveAllSSRCs();
 
-    if (const auto capture_id = m_devices.GetDefaultCapture(); capture_id.has_value()) {
-        m_capture_id = *capture_id;
-        m_capture_config.capture.pDeviceID = &m_capture_id;
+#ifdef WITH_RNNOISE
+    RNNoiseUninitialize();
+#endif
+}
 
-        if (auto code = ma_device_init(&m_context, &m_capture_config, &m_capture_device); code != MA_SUCCESS) {
-            spdlog::get("audio")->error("failed to initialize capture device (code: {})", static_cast<int>(code));
-            m_ok = false;
-            return;
-        }
+void AudioManager::OpenPlaybackDevice(const ma_device_id device_id) {
+    assert(!m_playback_device_ready && "Tried to open new playback device without closing the current one");
```

Comment:
i dont use assert anywhere else. would logging a warning not be sufficient ?

Reply:
I decided to use assert here, since calling Open/Close methods in invalid state would be an error in the code and not an expected behavior. If it's preferable to instead log it, I may do that.

Reply:
my understanding is assert is basically debug-only and ends up calling exit(). i think a message like this is also useful in release builds and even if it is triggered isnt a fatal-force-quit-tier occurrence. so imo i would just log it as an error or warning